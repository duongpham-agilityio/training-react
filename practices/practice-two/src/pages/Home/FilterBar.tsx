import { MouseEvent, ChangeEvent, memo, useCallback } from 'react';
import { Link } from 'react-router-dom';
import isEqual from 'react-fast-compare';

// Component
import { Button, Input } from 'components/commons';

// Styles
import homeStyles from 'pages/Home/index.module.css';

// Assets
import SearchIcon from 'assets/icons/search.svg';
import AddIcon from 'assets/icons/add.svg';

export interface SortOption {
  href: string;
  isActive: boolean;
  title: string;
}

export interface FilterBarProps {
  searchValue: string;
  sortOptions: SortOption[];
  onChangeSearch: (value: string) => void;
  onAddBook: (e: MouseEvent) => void;
}

const FilterBar = ({
  searchValue,
  sortOptions,
  onChangeSearch,
  onAddBook,
}: FilterBarProps) => {
  const changeSearchValue = useCallback(
    (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      onChangeSearch(event.target.value);
    },
    [onChangeSearch]
  );

  const renderOption = useCallback((option: SortOption) => {
    const { href, title, isActive } = option;

    return (
      <li key={title}>
        <Link
          to={href}
          className={`${homeStyles.navLink} ${
            isActive ? homeStyles.active : ''
          }`}
        >
          {title}
        </Link>
      </li>
    );
  }, []);

  return (
    <div className={homeStyles.navbar}>
      <Input
        className={homeStyles.search}
        value={searchValue}
        leftIcon={SearchIcon}
        placeholder="Search something..."
        onChange={changeSearchValue}
      />
      <ul className={homeStyles.navList}>{sortOptions.map(renderOption)}</ul>
      <Button
        label="Create"
        variant="primary"
        leftIcon={AddIcon}
        width="w-lg"
        border="b-lg"
        onClick={onAddBook}
      />
    </div>
  );
};

const areCompare = (prev: FilterBarProps, next: FilterBarProps): boolean =>
  isEqual(prev.sortOptions, next.sortOptions) &&
  isEqual(prev.searchValue, next.searchValue);

export default memo(FilterBar, areCompare);
