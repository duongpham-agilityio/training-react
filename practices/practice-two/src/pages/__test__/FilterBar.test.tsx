// Mocks
import 'components/__test__/mocks/image.test';

import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

// Components
import FilterBar, { FilterBarProps } from 'pages/Home/FilterBar';

jest.mock('react-fast-compare', () => ({
  default: jest.fn(),
}));

const setup = (props: FilterBarProps) =>
  render(<FilterBar {...props} />, {
    wrapper: BrowserRouter,
  });

const defaultProps: FilterBarProps = {
  searchValue: '',
  sortOptions: [
    {
      href: '#one',
      isActive: false,
      title: 'One',
    },
    {
      href: '#two',
      isActive: false,
      title: 'Two',
    },
  ],
  onAddBook: jest.fn(),
  onChangeSearch: jest.fn(),
};

describe('Filter Bar', () => {
  it('Match to snapshot', () => {
    const { container } = setup(defaultProps);

    expect(container).toMatchSnapshot();
  });

  it('Match to snapshot', () => {
    const { rerender } = setup(defaultProps);

    rerender(<FilterBar {...defaultProps} searchValue="html" />);
  });
});
