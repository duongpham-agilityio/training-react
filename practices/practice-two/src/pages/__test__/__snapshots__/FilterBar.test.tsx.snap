// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Bar Match to snapshot 1`] = `
<div>
  <div>
    <div
      class="undefined function search() { [native code] }"
    >
      <button
        aria-label="Search button"
        class="undefined undefined
  undefined undefined
  undefined "
        style="background-color: transparent; padding: 0px;"
        type="button"
      >
        <img
          alt="This is the left icon for the button"
          height="25px"
          src="assets/icons/search.svg"
          width="22px"
        />
      </button>
      <input
        placeholder="Search something..."
        type="text"
        value=""
      />
      <button
        aria-label="The right button"
        class="undefined undefined
  undefined undefined
  undefined "
        style="background-color: transparent; padding: 0px;"
        type="button"
      />
    </div>
    <ul>
      <li>
        <a
          class="undefined "
          href="/#one"
        >
          One
        </a>
      </li>
      <li>
        <a
          class="undefined "
          href="/#two"
        >
          Two
        </a>
      </li>
    </ul>
    <button
      class="undefined undefined
  undefined undefined
  undefined "
      type="button"
    >
      <img
        alt="This is the left icon for the button"
        height="25px"
        src="assets/icons/add.svg"
        width="22px"
      />
      Create
    </button>
  </div>
</div>
`;
